akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventConsumerServiceImpl.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.grpc.internal
+
+import akka.NotUsed
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.scaladsl.AskPattern._
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.grpc.scaladsl.Metadata
+import akka.persistence.EventWriter
+import akka.persistence.FilteredPayload
+import akka.projection.grpc.consumer.scaladsl.EventConsumerInterceptor
+import akka.projection.grpc.internal.proto.ConsumeEventIn
+import akka.projection.grpc.internal.proto.ConsumeEventOut
+import akka.projection.grpc.internal.proto.ConsumerEventAck
+import akka.projection.grpc.internal.proto.EventConsumerServicePowerApi
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import akka.util.Timeout
+import org.slf4j.LoggerFactory
+
+import java.net.URLEncoder
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.DurationInt
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ *
+ * gRPC push protocol service for the consuming side
+ */
+private[akka] object EventConsumerServiceImpl {
+
+  /**
+   *
+   * @param journalPluginId empty to use default
+   */
+  def directJournalConsumer(
+      journalPluginId: Option[String],
+      acceptedStreamIds: Set[String],
+      persistenceIdTransformer: String => String,
+      interceptor: Option[EventConsumerInterceptor])(implicit system: ActorSystem[_]): EventConsumerServiceImpl = {
+    // FIXME is this name unique, could we create multiple for the same journal? (we wouldn't be able to bind them to the same port)
+    val eventWriter = system.systemActorOf(
+      EventWriter(journalPluginId.getOrElse("")),
+      s"EventWriter-${URLEncoder.encode(journalPluginId.getOrElse("default"), ByteString.UTF_8)}")
+
+    new EventConsumerServiceImpl(eventWriter, acceptedStreamIds, persistenceIdTransformer, interceptor)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] final class EventConsumerServiceImpl(
+    eventWriter: ActorRef[EventWriter.Command],
+    acceptedStreamIds: Set[String],
+    persistenceIdTransformer: String => String,
```

Comment:
Yeah, I have been thinking about the user level API of it a bit. Something like the consumer `Transformation` API, which is EventEnvelope => payload, but we want to support transforming pids and maybe adding tags, so I think it would need to be its own API in that case (but I'd wish it was re-usable).

Reply:
Yes, maybe worth it's own api.